File: mercury/trager_factoring.m
%---------------------------------------------------------------------------%
% vim: ft=mercury ts=4 sw=4 et
%---------------------------------------------------------------------------%
% File: trager_factoring.m
% Main author: nyc
%
% Factoring polynomials over algebraic extension fields Q(α)
% using Trager's algorithm (1976).
%
% Given f(x) in Q(α)[x] where α is a root of an irreducible m(t) in Q[t],
% compute the norm N(x) = Res_t(m(t), f(x - s*t)) in Q[x] for suitable s,
% factor N over Q, and lift factors back via GCD.
%
%---------------------------------------------------------------------------%

:- module trager_factoring.
:- interface.

:- import_module extension.
:- import_module list.
:- import_module poly.
:- import_module rational.

%---------------------------------------------------------------------------%

    % Factor a polynomial over Q(α) into irreducible factors.
    % Returns a list of (factor, multiplicity) pairs.
    %
:- func factor_over_extension(ext_field(rational),
    poly(ext_elem(rational))) = list({poly(ext_elem(rational)), int}).

    % Factor a square-free polynomial over Q(α) into irreducible factors.
    %
:- func factor_sf_over_extension(ext_field(rational),
    poly(ext_elem(rational))) = list(poly(ext_elem(rational))).

    % Generalized: factor a square-free polynomial over K(α) given a
    % factoring function for K.
    %
:- func factor_sf_over_extension_k(
    (func(poly(K)) = list(poly(K))),
    ext_field(K),
    poly(ext_elem(K))) = list(poly(ext_elem(K))) <= field(K).

    % Compute the norm of f(x) in Q(α)[x]:
    % N(x) = Res_t(m(t), f_lifted(x, t)) in Q[x].
    %
:- func norm_poly(ext_field(rational),
    poly(ext_elem(rational))) = poly(rational).

    % Generalized norm over any base field K.
    %
:- func norm_poly_k(ext_field(K), poly(ext_elem(K))) = poly(K) <= field(K).

%---------------------------------------------------------------------------%
%---------------------------------------------------------------------------%

:- implementation.

:- use_module int.
:- import_module factoring.
:- import_module require.
:- import_module resultant.

%---------------------------------------------------------------------------%
% Factor over Q(α)
%---------------------------------------------------------------------------%

factor_over_extension(Field, F) = Result :-
    SFFactors = square_free(F),
    Result = list.condense(list.map(
        ( func(SFF) = list.map(
            ( func(H) = {H, sff_multiplicity(SFF)} ),
            factor_sf_over_extension(Field, sff_factor(SFF)))
        ), SFFactors)).

factor_sf_over_extension(Field, F) = Result :-
    D = degree(F),
    ( if int.'=<'(D, 0) then
        Result = []
    else if D = 1 then
        Result = [monic(F)]
    else
        Result = trager_factor(Field, F, 0)
    ).

:- func trager_factor(ext_field(rational),
    poly(ext_elem(rational)), int) = list(poly(ext_elem(rational))).

trager_factor(Field, F, S) = Result :-
    ( if int.'>'(S, 20) then
        Result = [F]
    else
        Alpha = generator(Field),
        SAlpha = ext_mul(embed(Field, rational(S, 1)), Alpha),
        FShifted = shift_poly(F, SAlpha),
        N = norm_poly(Field, FShifted),
        ( if is_square_free_q(N) then
            NFactors = factor_square_free(monic(N)),
            ( if int.'=<'(list.length(NFactors), 1) then
                Result = [monic(F)]
            else
                Result = lift_factors(Field, F, SAlpha, NFactors)
            )
        else
            Result = trager_factor(Field, F, int.'+'(S, 1))
        )
    ).

%---------------------------------------------------------------------------%
% Norm computation
%---------------------------------------------------------------------------%

norm_poly(Field, F) = Result :-
    DF = degree(F),
    DM = ext_degree(Field),
    ResultDeg = int.'*'(DF, DM),
    make_points(0, ResultDeg, Points),
    Values = list.map(( func(X0) = norm_at_point(Field, F, X0) ), Points),
    list.map_corresponding(( pred(P::in, V::in, PV::out) is det :-
        PV = {P, V}
    ), Points, Values, PVs),
    Result = lagrange_interpolate(PVs).

:- pred make_points(int::in, int::in,
    list(rational)::out) is det.

make_points(I, Max, Points) :-
    ( if int.'>'(I, Max) then
        Points = []
    else
        make_points(int.'+'(I, 1), Max, RestPoints),
        Points = [rational(I, 1) | RestPoints]
    ).

:- func norm_at_point(ext_field(rational),
    poly(ext_elem(rational)), rational) = rational.

norm_at_point(Field, F, X0) = Result :-
    M = gen_min_poly(Field),
    % f_lifted(x0, t) = Σ c_i(t) * x0^i
    Cs = coeffs(F),
    eval_lifted(Cs, X0, 0, zero_poly, FAtX0),
    Result = poly_resultant(M, FAtX0).

:- pred eval_lifted(list(ext_elem(rational))::in, rational::in,
    int::in, poly(rational)::in, poly(rational)::out) is det.

eval_lifted([], _, _, Acc, Acc).
eval_lifted([C | Cs], X0, I, Acc0, Acc) :-
    CP = elem_poly(C),
    X0I = rat_pow_q(X0, I),
    Term = scale(X0I, CP),
    Acc1 = add(Acc0, Term),
    eval_lifted(Cs, X0, int.'+'(I, 1), Acc1, Acc).

:- func rat_pow_q(rational, int) = rational.

rat_pow_q(X, N) = Result :-
    ( if N = 0 then
        Result = rational.one
    else if int.even(N) then
        Half = rat_pow_q(X, int.'//'(N, 2)),
        Result = rational.'*'(Half, Half)
    else
        Rest = rat_pow_q(X, int.'-'(N, 1)),
        Result = rational.'*'(X, Rest)
    ).

%---------------------------------------------------------------------------%
% Square-free test
%---------------------------------------------------------------------------%

:- pred is_square_free_q(poly(rational)::in) is semidet.

is_square_free_q(P) :-
    P1 = diff(P),
    G = gcd(P, P1),
    int.'=<'(degree(G), 0).

%---------------------------------------------------------------------------%
% Lifting
%---------------------------------------------------------------------------%

:- func lift_factors(ext_field(rational),
    poly(ext_elem(rational)),
    ext_elem(rational),
    list(poly(rational))) = list(poly(ext_elem(rational))).

lift_factors(Field, F, SAlpha, QFactors) =
    lift_factors_loop(Field, F, SAlpha, QFactors).

:- func lift_factors_loop(ext_field(rational),
    poly(ext_elem(rational)),
    ext_elem(rational),
    list(poly(rational))) = list(poly(ext_elem(rational))).

lift_factors_loop(_, _, _, []) = [].
lift_factors_loop(Field, Remaining, SAlpha, [G | Gs]) = Result :-
    ( if int.'=<'(degree(Remaining), 0) then
        Result = []
    else
        GLifted = lift_q_poly(Field, G),
        GUnshifted = shift_poly(GLifted, ext_neg(SAlpha)),
        H = gcd_poly_ext(Remaining, GUnshifted),
        ( if int.'>'(degree(H), 0) then
            div_mod(Remaining, H, Q, _),
            Result = [monic(H) |
                lift_factors_loop(Field, Q, SAlpha, Gs)]
        else
            Result = lift_factors_loop(Field, Remaining, SAlpha, Gs)
        )
    ).

:- func lift_q_poly(ext_field(rational), poly(rational))
    = poly(ext_elem(rational)).

lift_q_poly(Field, P) =
    poly(list.map(embed(Field), coeffs(P))).

:- func gcd_poly_ext(poly(ext_elem(rational)),
    poly(ext_elem(rational))) = poly(ext_elem(rational)).

gcd_poly_ext(A, B) = Result :-
    ( if is_zero_poly_ext(B) then
        Result = monic(A)
    else
        div_mod(A, B, _, R),
        Result = gcd_poly_ext(B, R)
    ).

:- pred is_zero_poly_ext(poly(ext_elem(rational))::in) is semidet.

is_zero_poly_ext(P) :-
    list.all_true(( pred(C::in) is semidet :-
        ring_is_zero(C)
    ), coeffs(P)).

%---------------------------------------------------------------------------%
% Generalized Trager over any K
%---------------------------------------------------------------------------%

factor_sf_over_extension_k(FactorBaseK, Field, F) = Result :-
    D = degree(F),
    ( if int.'=<'(D, 0) then
        Result = []
    else if D = 1 then
        Result = [monic(F)]
    else
        Result = trager_factor_k(FactorBaseK, Field, F, 0)
    ).

:- func trager_factor_k(
    (func(poly(K)) = list(poly(K))),
    ext_field(K),
    poly(ext_elem(K)),
    int) = list(poly(ext_elem(K))) <= field(K).

trager_factor_k(FactorBaseK, Field, F, S) = Result :-
    ( if int.'>'(S, 20) then
        Result = [F]
    else
        Alpha = generator(Field),
        SAlpha = ext_mul(
            embed(Field, ring_from_int(S)),
            Alpha),
        FShifted = shift_poly(F, SAlpha),
        N = norm_poly_k(Field, FShifted),
        ( if is_square_free_k(N) then
            NFactors = FactorBaseK(monic(N)),
            ( if int.'=<'(list.length(NFactors), 1) then
                Result = [monic(F)]
            else
                Result = lift_factors_k(Field, F, SAlpha, NFactors)
            )
        else
            Result = trager_factor_k(FactorBaseK, Field, F,
                int.'+'(S, 1))
        )
    ).

norm_poly_k(Field, F) = Result :-
    DF = degree(F),
    DM = ext_degree(Field),
    ResultDeg = int.'*'(DF, DM),
    M = gen_min_poly(Field),
    make_points_k(0, ResultDeg, Points),
    Values = list.map(( func(X0) = R :-
        FAtX0 = eval_poly_ext_k(F, X0),
        R = poly_resultant(M, FAtX0)
    ), Points),
    list.map_corresponding(( pred(P::in, V::in, PV::out) is det :-
        PV = {P, V}
    ), Points, Values, PVs),
    Result = lagrange_interpolate(PVs).

:- pred make_points_k(int::in, int::in,
    list(K)::out) is det <= field(K).

make_points_k(I, Max, Points) :-
    ( if int.'>'(I, Max) then
        Points = []
    else
        make_points_k(int.'+'(I, 1), Max, RestPoints),
        Points = [ring_from_int(I) | RestPoints]
    ).

:- func eval_poly_ext_k(poly(ext_elem(K)), K) = poly(K) <= field(K).

eval_poly_ext_k(F, X0) = Result :-
    Cs = coeffs(F),
    eval_lifted_k(Cs, X0, 0, zero_poly, Result).

:- pred eval_lifted_k(list(ext_elem(K))::in, K::in,
    int::in, poly(K)::in, poly(K)::out) is det <= field(K).

eval_lifted_k([], _, _, Acc, Acc).
eval_lifted_k([C | Cs], X0, I, Acc0, Acc) :-
    CP = elem_poly(C),
    X0I = pow_k(X0, I),
    Term = scale(X0I, CP),
    Acc1 = add(Acc0, Term),
    eval_lifted_k(Cs, X0, int.'+'(I, 1), Acc1, Acc).

:- func pow_k(K, int) = K <= field(K).

pow_k(X, N) = Result :-
    ( if N = 0 then
        Result = ring_one
    else if int.even(N) then
        Half = pow_k(X, int.'//'(N, 2)),
        Result = ring_mul(Half, Half)
    else
        Result = ring_mul(X, pow_k(X, int.'-'(N, 1)))
    ).

:- pred is_square_free_k(poly(K)::in) is semidet <= field(K).

is_square_free_k(P) :-
    P1 = diff(P),
    G = gcd(P, P1),
    int.'=<'(degree(G), 0).

:- func lift_factors_k(ext_field(K),
    poly(ext_elem(K)),
    ext_elem(K),
    list(poly(K))) = list(poly(ext_elem(K))) <= field(K).

lift_factors_k(_, _, _, []) = [].
lift_factors_k(Field, Remaining, SAlpha, [G | Gs]) = Result :-
    ( if int.'=<'(degree(Remaining), 0) then
        Result = []
    else
        GLifted = lift_k_poly(Field, G),
        GUnshifted = shift_poly(GLifted, ext_neg(SAlpha)),
        H = gcd_poly_generic(Remaining, GUnshifted),
        ( if int.'>'(degree(H), 0) then
            div_mod(Remaining, H, Q, _),
            Result = [monic(H) |
                lift_factors_k(Field, Q, SAlpha, Gs)]
        else
            Result = lift_factors_k(Field, Remaining, SAlpha, Gs)
        )
    ).

:- func lift_k_poly(ext_field(K), poly(K)) = poly(ext_elem(K)) <= field(K).

lift_k_poly(Field, P) =
    poly(list.map(embed(Field), coeffs(P))).

:- func gcd_poly_generic(poly(ext_elem(K)),
    poly(ext_elem(K))) = poly(ext_elem(K)) <= field(K).

gcd_poly_generic(A, B) = Result :-
    ( if is_zero_poly_ext_k(B) then
        Result = monic(A)
    else
        div_mod(A, B, _, R),
        Result = gcd_poly_generic(B, R)
    ).

:- pred is_zero_poly_ext_k(poly(ext_elem(K))::in) is semidet <= field(K).

is_zero_poly_ext_k(P) :-
    list.all_true(( pred(C::in) is semidet :-
        ring_is_zero(C)
    ), coeffs(P)).

%---------------------------------------------------------------------------%
:- end_module trager_factoring.
%---------------------------------------------------------------------------%

File: haskell/algebraic-polynomials/src/Math/Polynomial/TragerFactoring.hs
-- |
-- Module      : Math.Polynomial.TragerFactoring
-- Description : Factoring polynomials over algebraic extension fields via Trager's algorithm
-- Stability   : experimental
--
-- Implementation of Trager's algorithm (1976) for factoring polynomials
-- over simple algebraic extension fields \(K(\alpha)\).
--
-- __Algorithm outline:__
--
-- Given \(f(x) \in K(\alpha)[x]\) where \(\alpha\) is a root of an
-- irreducible \(m(t) \in K[t]\):
--
-- 1. Compute the norm \(N(x) = \text{Res}_t(m(t), f(x - s \cdot t)) \in K[x]\)
--    for a suitable shift parameter \(s\) (chosen so that \(N\) is square-free).
-- 2. Factor \(N(x)\) over the base field \(K\).
-- 3. Lift factors back to \(K(\alpha)[x]\) via GCD:
--    for each factor \(g_i\) of \(N\), compute
--    \(\gcd(f(x), g_i(x + s \cdot \alpha))\) in \(K(\alpha)[x]\).
--
-- Two variants are provided:
--
-- * 'factorSFOverExtension': factors over \(\mathbb{Q}(\alpha)\),
--   using 'Math.Polynomial.Factoring.factorSquareFree' for the base case.
-- * 'factorSFOverExtensionK': generalized to arbitrary base field \(K\),
--   taking a base-field factoring function as a parameter (enabling
--   iterated extension towers like \(\mathbb{Q}(\alpha)(\beta)\)).
--
-- Reference: B. M. Trager, \"Algebraic Factoring and Rational Function
-- Integration\", /Proceedings of SYMSAC '76/, pp. 219--226, ACM, 1976.
module Math.Polynomial.TragerFactoring
  ( factorOverExtension
  , factorSFOverExtension
  , factorSFOverExtensionK
  , normPoly
  , normPolyK
  ) where

import Math.Polynomial.Univariate
import Math.Polynomial.Factoring (factorSquareFree)
import Math.Polynomial.Resultant (polyResultant, lagrangeInterpolate, shiftPoly)
import Math.Field.Extension

-- | Factor a polynomial over \(\mathbb{Q}(\alpha)\) into irreducible factors.
--
-- Returns a list of @(factor, multiplicity)@ pairs. Uses square-free
-- factorisation followed by 'factorSFOverExtension' on each component.
factorOverExtension :: ExtField Rational
                    -> Poly (ExtElem Rational)
                    -> [(Poly (ExtElem Rational), Int)]
factorOverExtension field f =
  let sfFactors = squareFreeExt field f
  in concatMap (\(g, m) -> map (\h -> (h, m)) (factorSFOverExtension field g)) sfFactors

-- | Factor a square-free polynomial over \(\mathbb{Q}(\alpha)\) into
-- irreducible factors using Trager's algorithm.
--
-- The shift parameter \(s\) is incremented from 0 until the norm
-- polynomial is square-free. If no factorisation is found after 20
-- shifts, the polynomial is returned as-is (assumed irreducible).
--
-- Returns a list of monic irreducible factors.
factorSFOverExtension :: ExtField Rational
                      -> Poly (ExtElem Rational)
                      -> [Poly (ExtElem Rational)]
factorSFOverExtension field f
  | degree f <= 0 = []
  | degree f == 1 = [monicPoly f]
  | otherwise = tragerFactor field f 0

tragerFactor :: ExtField Rational
             -> Poly (ExtElem Rational)
             -> Int  -- shift parameter s
             -> [Poly (ExtElem Rational)]
tragerFactor field f s
  | s > 20 = [f]  -- give up, return as-is
  | otherwise =
      let alpha = generator field
          -- f_shifted(x) = f(x + s*alpha) -- shift so norm is square-free
          sAlpha = embed field (fromIntegral s) * alpha
          fShifted = shiftPoly f sAlpha
          -- Compute norm: N(x) = Res_t(m(t), fShifted(x, t))
          n = normPoly field fShifted
      in if not (isSquareFreeQ n)
         then tragerFactor field f (s + 1)
         else
           -- Factor N(x) over Q
           let nFactors = factorSquareFree (monicPoly n)
           in if length nFactors <= 1
              then [monicPoly f]  -- f is irreducible over Q(alpha)
              else
                -- Lift each factor back
                let lifted = liftFactors field f sAlpha nFactors
                in lifted

-- | Compute the norm of \(f(x) \in \mathbb{Q}(\alpha)[x]\):
-- \(N(x) = \text{Res}_t(m(t), f_{\text{lifted}}(x, t)) \in \mathbb{Q}[x]\).
--
-- The lifted polynomial \(f_{\text{lifted}}(x, t)\) replaces each
-- coefficient \(c \in \mathbb{Q}(\alpha)\) (represented as a polynomial
-- in \(\alpha\)) with that polynomial in the variable \(t\).
--
-- Computed via evaluation at \(\deg f \cdot \deg m + 1\) rational points
-- followed by Lagrange interpolation.
normPoly :: ExtField Rational -> Poly (ExtElem Rational) -> Poly Rational
normPoly field f =
  let df = degree f
      dm = extDegree field
      resultDeg = df * dm
      -- Evaluate at 0, 1, 2, ..., resultDeg
      points = [fromIntegral i | i <- [0..resultDeg]]
      values = [normAtPoint field f x0 | x0 <- points]
  in lagrangeInterpolate (zip points values)

-- | Evaluate the norm at a specific rational point \(x_0\):
-- \(\text{Res}_t(m(t), f_{\text{lifted}}(x_0, t))\).
normAtPoint :: ExtField Rational -> Poly (ExtElem Rational) -> Rational -> Rational
normAtPoint field f x0 =
  let m = genMinPoly field
      -- f(x) = c_0 + c_1*x + c_2*x^2 + ...
      -- f_lifted(x0, t) = c_0(t) + c_1(t)*x0 + c_2(t)*x0^2 + ...
      -- where c_i(t) is the polynomial representation of the i-th coefficient.
      coeffs = map elemPoly (unPoly f)
      -- f_lifted(x0, t) = sum c_i(t) * x0^i
      fAtX0 = case coeffs of
                [] -> zeroPoly
                _  -> foldl addPoly zeroPoly
                        [scalePoly (x0 ^ i) ci | (i, ci) <- zip [0 :: Int ..] coeffs]
  in polyResultant m fAtX0

-- | Lift Q-factors back to Q(alpha)-factors via GCD.
liftFactors :: ExtField Rational
            -> Poly (ExtElem Rational)  -- f(x)
            -> ExtElem Rational         -- s*alpha (the shift)
            -> [Poly Rational]          -- factors of N(x) over Q
            -> [Poly (ExtElem Rational)]
liftFactors field f sAlpha qFactors = go f qFactors
  where
    go _ [] = []
    go remaining (g : gs)
      | degree remaining <= 0 = []
      | otherwise =
          -- g_i(x - s*alpha) as a polynomial in Q(alpha)[x]
          let gLifted = liftQPoly field g
              gUnshifted = shiftPoly gLifted (negate sAlpha)
              -- gcd(remaining, gUnshifted) in Q(alpha)[x]
              h = gcdPolyExt field remaining gUnshifted
          in if degree h > 0
             then let (q, _) = divModPolyExt field remaining h
                  in monicPoly h : go q gs
             else go remaining gs

-- | Lift a \(\mathbb{Q}[x]\) polynomial to \(\mathbb{Q}(\alpha)[x]\) by
-- embedding coefficients.
liftQPoly :: ExtField Rational -> Poly Rational -> Poly (ExtElem Rational)
liftQPoly field (Poly cs) = Poly (map (embed field) cs)

-- | Check if a rational polynomial is square-free (has no repeated roots).
isSquareFreeQ :: Poly Rational -> Bool
isSquareFreeQ p =
  let p' = diffPoly p
      g = gcdPoly p p'
  in degree g <= 0

-- | GCD of two polynomials over Q(alpha).
gcdPolyExt :: ExtField Rational
           -> Poly (ExtElem Rational)
           -> Poly (ExtElem Rational)
           -> Poly (ExtElem Rational)
gcdPolyExt field a b
  | isZeroPolyExt b = monicPoly a
  | otherwise = gcdPolyExt field b (snd $ divModPolyExt field a b)

-- | Division of polynomials over Q(alpha).
divModPolyExt :: ExtField Rational
              -> Poly (ExtElem Rational)
              -> Poly (ExtElem Rational)
              -> (Poly (ExtElem Rational), Poly (ExtElem Rational))
divModPolyExt _ f g
  -- Use the generic divModPoly, which works because ExtElem has Fractional
  = divModPoly f g

-- | Check if a polynomial over Q(alpha) is zero.
isZeroPolyExt :: Poly (ExtElem Rational) -> Bool
isZeroPolyExt (Poly cs) = all (== 0) cs

-- | Square-free factorization over Q(alpha).
squareFreeExt :: ExtField Rational
              -> Poly (ExtElem Rational)
              -> [(Poly (ExtElem Rational), Int)]
squareFreeExt _ = squareFree

-- | Evaluate \(f(x_0)\) where \(f\) has 'ExtElem' coefficients and
-- \(x_0 \in K\), returning a polynomial in \(K[t]\) (the representation
-- of the result as an element of \(K(\alpha)\)).
evalPolyExtK :: (Eq k, Fractional k) => Poly (ExtElem k) -> k -> Poly k
evalPolyExtK (Poly []) _ = zeroPoly
evalPolyExtK (Poly cs) x0 =
  foldl (\acc (i, ExtElem p _) ->
           addPoly acc (scalePoly (x0 ^ i) p))
        zeroPoly
        (zip [0 :: Int ..] cs)

-- | Compute the norm of \(f(x) \in K(\alpha)[x]\) over any base field \(K\):
-- \(N(x) = \text{Res}_t(m(t), f_{\text{lifted}}(x, t)) \in K[x]\).
--
-- Generalizes 'normPoly' from \(\mathbb{Q}\) to arbitrary coefficient fields.
normPolyK :: (Eq k, Fractional k) => ExtField k -> Poly (ExtElem k) -> Poly k
normPolyK field f =
  let df = degree f
      dm = extDegree field
      resultDeg = df * dm
      m = genMinPoly field
      points = [fromInteger i | i <- [0..fromIntegral resultDeg]]
      values = [polyResultant m (evalPolyExtK f x0) | (x0 :: k) <- points]
  in lagrangeInterpolate (zip points values)

-- | Factor a square-free polynomial over \(K(\alpha)\) given a factoring
-- function for the base field \(K\).
--
-- This is the generalized Trager algorithm. The @factorBaseK@ parameter
-- provides factoring over the base field (e.g., 'factorSquareFree' for
-- \(\mathbb{Q}\), or @factorSFOverExtension field1@ for
-- \(K = \mathbb{Q}(\alpha_1)\)).
--
-- This enables factoring over field towers like
-- \(\mathbb{Q}(\alpha_1)(\alpha_2)\) by nesting.
factorSFOverExtensionK :: (Eq k, Fractional k)
                       => (Poly k -> [Poly k])  -- ^ Factor over base field K
                       -> ExtField k
                       -> Poly (ExtElem k)
                       -> [Poly (ExtElem k)]
factorSFOverExtensionK factorBaseK field f
  | degree f <= 0 = []
  | degree f == 1 = [monicPoly f]
  | otherwise = tragerFactorK factorBaseK field f 0

tragerFactorK :: (Eq k, Fractional k)
              => (Poly k -> [Poly k])
              -> ExtField k
              -> Poly (ExtElem k)
              -> Int
              -> [Poly (ExtElem k)]
tragerFactorK factorBaseK field f s
  | s > 20 = [f]
  | otherwise =
      let alpha = generator field
          sAlpha = embed field (fromInteger (fromIntegral s)) * alpha
          fShifted = shiftPoly f sAlpha
          n = normPolyK field fShifted
      in if not (isSquareFreeK n)
         then tragerFactorK factorBaseK field f (s + 1)
         else
           let nFactors = factorBaseK (monicPoly n)
           in if length nFactors <= 1
              then [monicPoly f]
              else liftFactorsK field f sAlpha nFactors

-- | Check if a polynomial over @k@ is square-free.
isSquareFreeK :: (Eq k, Fractional k) => Poly k -> Bool
isSquareFreeK p =
  let p' = diffPoly p
      g = gcdPolyK p p'
  in degree g <= 0

-- | GCD over any field.
gcdPolyK :: (Eq k, Fractional k) => Poly k -> Poly k -> Poly k
gcdPolyK a b
  | isZeroPoly b = monicPoly a
  | otherwise = gcdPolyK b (snd $ divModPoly a b)
  where
    isZeroPoly (Poly cs) = all (== 0) cs

-- | Lift base-field factors back to K(alpha)-factors via GCD (generalized).
liftFactorsK :: (Eq k, Fractional k)
             => ExtField k
             -> Poly (ExtElem k)
             -> ExtElem k
             -> [Poly k]
             -> [Poly (ExtElem k)]
liftFactorsK field f sAlpha kFactors = go f kFactors
  where
    go _ [] = []
    go remaining (g : gs)
      | degree remaining <= 0 = []
      | otherwise =
          let gLifted = liftKPoly field g
              gUnshifted = shiftPoly gLifted (negate sAlpha)
              h = gcdPolyGeneric remaining gUnshifted
          in if degree h > 0
             then let (q, _) = divModPoly remaining h
                  in monicPoly h : go q gs
             else go remaining gs

-- | GCD of two polynomials (generic, for ExtElem k).
gcdPolyGeneric :: (Eq a, Fractional a) => Poly a -> Poly a -> Poly a
gcdPolyGeneric a b
  | all (== 0) (unPoly b) = monicPoly a
  | otherwise = gcdPolyGeneric b (snd $ divModPoly a b)

-- | Lift a @K[x]@ polynomial to @K(alpha)[x]@ by embedding coefficients.
liftKPoly :: (Eq k, Num k) => ExtField k -> Poly k -> Poly (ExtElem k)
liftKPoly field (Poly cs) = Poly (map (embed field) cs)
